oap-storage-mongo/src/main/java/oap/storage/mongo/MongoPersistence.java

```diff
@@ -24,5 +24,177 @@
+import oap.concurrent.scheduler.Scheduled;
+import oap.concurrent.scheduler.Scheduler;
+import oap.reflect.TypeRef;
+import oap.storage.MemoryStorage;
+import oap.storage.Metadata;
+import oap.storage.Storage;
+import org.bson.codecs.configuration.CodecRegistries;
+import org.bson.codecs.configuration.CodecRegistry;
+
+import java.io.Closeable;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import static com.mongodb.client.model.Filters.eq;
+
+/**
+ * Mongo Persistence storage (similar to {@link oap.storage.DirectoryPersistence})
+ *
+ * @param <T> Type of Metadata
+ */
+@Slf4j
+@ThreadSafe
+public class MongoPersistence<T> implements Closeable, Storage.DataListener<T> {
+
+    public static final ReplaceOptions REPLACE_OPTIONS_UPSERT = new ReplaceOptions().upsert( true );
+    public final MongoCollection<Metadata<T>> collection;
+    private long delay;
+    private final Lock lock = new ReentrantLock();
+    private MemoryStorage<T> storage;
+    private PeriodicScheduled scheduled;
+
+    public MongoPersistence( MongoClientWrapper mongoClient,
+                             String table,
+                             long delay,
+                             MemoryStorage<T> storage ) {
+        this.delay = delay;
+        this.storage = storage;
+
+        TypeRef<Metadata<T>> ref = new TypeRef<>() {};
+
+        CodecRegistry codecRegistry = CodecRegistries.fromRegistries(
+            CodecRegistries.fromCodecs( new JsonCodec<>( ref, m -> this.storage.identifier.get( m.object ) ) ),
+            mongoClient.database.getCodecRegistry()
+        );
+
+        this.collection = mongoClient.database
+            .getCollection( table, ref.clazz() )
+            .withCodecRegistry( codecRegistry );
+    }
+
+    public void start() {
+        this.load();
+        this.scheduled = Scheduler.scheduleWithFixedDelay( getClass(), delay, this::fsync );
+        this.storage.addDataListener( this );
+    }
+
+    /**
+     * Load all {@link MongoCollection} documents to the {@link MemoryStorage#data}
+     */
+    private void load() {
+        MongoNamespace namespace = collection.getNamespace();
+        log.debug( "loading data from {}", namespace );
+        Threads.synchronously( lock, () -> {
+            final Consumer<Metadata<T>> cons = metadata -> {
+                var id = storage.identifier.get( metadata.object );
+                storage.data.put( id, metadata );
+            };
+            log.info( "Load {} documents from [{}] Mongo namespace", collection.countDocuments(), namespace );
+            collection.find().forEach( cons );
+        } );
+        log.info( storage.data.size() + " object(s) loaded." );
+    }
+
+    @Override
+    public void updated( T object, boolean added ) {
+        Threads.synchronously( lock, () -> {
```

Comment:
This implementation makes all work serializable and synchronized for one lock. In DirectoryPErsistance that lock was only ment to sync start/fsync/stop to prevent concurrent stopping and fsyncing. Not operations.  HAve a look at synchronization on id concept.